src/operator/tensor/indexing_op.h

```diff
@@ -317,23 +380,21 @@ void EmbeddingOpBackward(const nnvm::NodeAttrs& attrs,
 }
 
 template<int req>
-struct EmbeddingBackwardRsp {
+struct EmbeddingBackwardDns {
   template<typename DType, typename IType>
   // each thread i is responsible for target gradient row ids in [segment_start, segment_end)
-  MSHADOW_XINLINE static void Map(int i, const size_t width, IType* dst_idx, DType* dst_val,
+  MSHADOW_XINLINE static void Map(int i, const index_t num_cols, DType* dst_val,
                                   const IType* idx, const size_t num_idx, const DType* src,
-                                  const size_t segment_len, const size_t num_rows) {
+                                  const int segment_len, const size_t num_rows) {
     auto req_type = req;
-    size_t segment_start = i * segment_len;
-    size_t segment_end = (i + 1) * segment_len;
+    auto segment_start = i * segment_len;
```

Comment:
segment_start might be greater than num_rows if there are more threads than num_rows. Do we need a check here?